src/Compilers/CSharp/Test/Semantic/SourceGeneration/GeneratorDriverTests.cs

```diff
@@ -40,7 +40,74 @@ class C { }
 
             Assert.Single(outputCompilation.SyntaxTrees);
             Assert.Equal(compilation, outputCompilation);
+        }
+
+        [Fact]
+        public void Generator_Is_Intialized_Before_Running()
+        {
+            var source = @"
+class C { }
+";
+            var parseOptions = TestOptions.Regular;
+            Compilation compilation = CreateCompilationWithMscorlib45(source, options: TestOptions.DebugDll, parseOptions: parseOptions);
+            compilation.VerifyDiagnostics();
+
+            Assert.Single(compilation.SyntaxTrees);
+
+            int initCount = 0, executeCount = 0;
+            var generator = new CallbackGenerator((ic) => initCount++, (sgc) => executeCount++);
+
+            GeneratorDriver driver = new CSharpGeneratorDriver(compilation, parseOptions, ImmutableArray.Create<ISourceGenerator>(generator), ImmutableArray<AdditionalText>.Empty);
+            driver.RunFullGeneration(compilation, out var outputCompilation);
+
+            Assert.Equal(1, initCount);
+            Assert.Equal(1, executeCount);
+        }
+
+        [Fact]
+        public void Generator_Is_Not_Initialized_If_Not_Run()
+        {
+            var source = @"
+class C { }
+";
+            var parseOptions = TestOptions.Regular;
+            Compilation compilation = CreateCompilationWithMscorlib45(source, options: TestOptions.DebugDll, parseOptions: parseOptions);
+            compilation.VerifyDiagnostics();
+
+            Assert.Single(compilation.SyntaxTrees);
+
+            int initCount = 0, executeCount = 0;
+            var generator = new CallbackGenerator((ic) => initCount++, (sgc) => executeCount++);
+
+            GeneratorDriver driver = new CSharpGeneratorDriver(compilation, parseOptions, ImmutableArray.Create<ISourceGenerator>(generator), ImmutableArray<AdditionalText>.Empty);
+
+            Assert.Equal(0, initCount);
+            Assert.Equal(0, executeCount);
+        }
+
+        [Fact]
+        public void Generator_Is_Only_Initialized_Once()
+        {
+            var source = @"
+class C { }
+";
+            var parseOptions = TestOptions.Regular;
+            Compilation compilation = CreateCompilationWithMscorlib45(source, options: TestOptions.DebugDll, parseOptions: parseOptions);
+            compilation.VerifyDiagnostics();
+
+            Assert.Single(compilation.SyntaxTrees);
 
+            int initCount = 0, executeCount = 0;
+            var generator = new CallbackGenerator((ic) => initCount++, (sgc) => executeCount++);
+            var generator2 = new SingleFileTestGenerator("public class C { }");
```

Comment:
>generator2 [](start = 16, length = 10)

Does `generator2` affect this test?

Reply:
It's subtle, but it did: without adding source you end up not executing. I modified the callback generator to do the same thing.

---
In reply to: [392539551](https://github.com/dotnet/roslyn/pull/42359#discussion_r392539551) [](ancestors = 392539551)